feat(kubernetes/v2): Converts CopyToClipboard to React Component

DarrenN opened the pull request:
**features:**

- Removes original Angular component in `core/utils/clipboard/copyToClipboard.tsx`
- Replaces with a React component with similar behavior (`core/utils/clipboard/CopyToClipboars.tsx`)
- Replaces instance of Angular component in `executions/Execution.tsx`
- Replaces instance of Angular component in `ManifestStatus.tsx`

![screen recording 2019-01-25 at 10 54 am](https://user-images.githubusercontent.com/215030/51756720-adcfce00-208f-11e9-9486-99fc8ad00e22.gif)

DarrenN commented:
@christopherthielen howdy! First pass at replacing the Angular clipboard copy component with a React component. Added the ReactGA event handling as well.

christopherthielen commented on app/scripts/modules/core/src/pipeline/executions/execution/Execution.tsx:
```diff
@@ -13,11 +13,13 @@ import { IExecutionViewState, IPipelineGraphNode } from 'core/pipeline/config/gr
 import { ISortFilter } from 'core/filterModel';
 import { ExecutionState } from 'core/state';
 
+import { CopyToClipboard } from '@spinnaker/core';
```
```suggestion
import { CopyToClipboard } from 'core/utils';
```
The import rules are unfortunately confusing.   When inside the `core` module, we import from the `core/[section]` alias.  When inside a non-core module (such as `kubernetes` or `amazon`) we import from `@spinnaker/core`

christopherthielen commented on app/scripts/modules/core/src/utils/clipboard/CopyToClipboard.tsx:
```diff
@@ -1,5 +1,181 @@
+import * as React from 'react';
+import * as ReactGA from 'react-ga';
+import { OverlayTrigger, Tooltip } from 'react-bootstrap';
+import { PositionProperty } from 'csstype';
```
Nice, I didn't know this was a thing.

christopherthielen commented on app/scripts/modules/core/src/utils/clipboard/CopyToClipboard.tsx:
```diff
@@ -1,5 +1,181 @@
+  overflow?: string;
+  position?: PositionProperty;
+}
+
```
```ts
interface ICopyToClipboardState {
    tooltipCopy: boolean;
    inputWidth: number | 'auto';
};
```

christopherthielen commented on app/scripts/modules/core/src/utils/clipboard/CopyToClipboard.tsx:
```diff
@@ -1,5 +1,181 @@
+ * since the text is placed in an invisible input its very easy to select
+ * if the copy fails.
+ */
+export class CopyToClipboard extends React.Component<ICopyToClipboardProps> {
```
```suggestion
export class CopyToClipboard extends React.Component<ICopyToClipboardProps, ICopyToClipboardState> {
```

christopherthielen commented on app/scripts/modules/core/src/utils/clipboard/copyToClipboard.component.ts:
```diff
@@ -1,70 +0,0 @@
-import { IController, IComponentOptions, IScope, ITimeoutService, module } from 'angular';
```
The angular component is still being used in many places.  We can't delete this file unless we make a `react2angular` adapter as well.  Because the input props should match (between the react component and the angular component), I recommend renaming the `value` prop back to `text`.

The adapter code can go in `app/scripts/modules/core/src/utils/clipboard/CopyToClipboard.component.ts` and would look like:
```
import { module } from 'angular';
import { react2angular } from 'react2angular';

import { CopyToClipboard } from './CopyToClipboard';

export const COPY_TO_CLIPBOARD_COMPONENT = 'spinnaker.core.utils.clipboard.copyToClipboard';
module(COPY_TO_CLIPBOARD_COMPONENT, []).component(
  'copyToClipboard',
  react2angular(CopyToClipboard, ['text', 'toolTip', 'analyticsLabel']),
);
```

DarrenN commented:
@christopherthielen i updated with all of the suggestions and changes the props to match the existing Angular component. I replaced the Angular component, but am not getting a large and cryptic Angular error in the unit tests. I've never worked with Angular, so I really don't know what's happening there.

christopherthielen commented:
@DarrenN I'll pull it down and take a look

DarrenN commented:
Hmmm it seems like it passed in CI? Interesting, a particular test blows up when I run them locally.

```
FAILED TESTS:
  Controller: Instance Archetype Selector
    ✖ should select a profile, change it, then unselect it
      Chrome 71.0.3578 (Mac OS X 10.14.2)
    Error: [$injector:modulerr] Failed to instantiate module undefined due to:
    Error: [ng:areq] Argument 'module' is not a function, got undefined

    https://errors.angularjs.org/1.6.10/$injector/modulerr?p0=undefined&p1=Error%3A%20%5Bng%3Aareq%5D%20Argument%20'module'%20is%20not%20a%20function%2C%20got%20undefined%0Ahttps%3A%2F%2Ferrors.angularjs.org%2F1.6.10%2Fng%2Fareq%3Fp0%3Dmodule%26p1%3Dnot%2520a%2520function%252C%2520got%2520undefined%0A%20%20%20%20at%20eval%20(webpack%3A%2F%2F%2F.%2Fnode_modules%2Fangular%2Fangular.js%3F%3A125%3A12)%0A%20%20%20%20at%20assertArg%20(webpack%3A%2F%2F%2F.%2Fnode_modules%2Fangular%2Fangular.j ...
```

DarrenN commented:
@christopherthielen awesome! Thanks for your help/guidance on this.

christopherthielen commented:
@DarrenN this needs to be squashed and rebased, then it should be ready to merge.

DarrenN commented:
So, unclear what went down on the squash here.

DarrenN commented:
@christopherthielen yeah this squash went to hell so just started over w/ cherry-pick and made a new one in https://github.com/spinnaker/deck/pull/6451
